### ISSUE TITLE
Issues with Qt 6.8

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#8242]
Qt 6.8 Beta 1 [was released]. Doing the usual dance to get wheels:

```
mkdir wheels-68
cd wheels-68
python3 -m venv .venv
.venv/bin/pip install pyqt-builder
.venv/bin/pip download --no-deps PyQt6 PyQtWebEngine
.venv/bin/pyqt-bundle --verbose --ignore-missing --qt-dir ~/proj/qt/bin6/6.8.0/gcc_64 PyQt6-6.7.0-1-cp38-abi3-manylinux_2_28_x86_64.whl
.venv/bin/pyqt-bundle --verbose --ignore-missing --qt-dir ~/proj/qt/bin6/6.8.0/gcc_64 PyQt6_WebEngine-6.7.0-cp38-abi3-manylinux_2_28_x86_64.whl
rm PyQt6-6.7.0-1-cp38-abi3-manylinux_2_28_x86_64.whl PyQt6_WebEngine-6.7.0-cp38-abi3-manylinux_2_28_x86_64.whl
cd ../git
python3 scripts/mkvenv.py --dev --venv-dir .venv-qt68 --pyqt-type wheels --pyqt-wheels-dir ../wheels-68
```

Once again leads to a missing symbol....

```
==================== Running Qt smoke test ====================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/florian/proj/qutebrowser/git/qutebrowser/qt/widgets.py", line 24, in <module>
    from PyQt6.QtWidgets import *
ImportError: /home/florian/proj/qutebrowser/git/.venv-qt68/lib/python3.12/site-packages/PyQt6/QtGui.abi3.so: undefined symbol: _ZN11QPageLayout15setBottomMarginEd, version Qt_6
```

**Comment by The-Compiler** [comment]
# Broken ABI

[[QTBUG-126454] [REG 6.7 -> 6.8] Broken ABI due to missing QPageLayout::setBottomMargin(qreal) - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-126454)

**Comment by The-Compiler** [comment]
# Missing security patch version in our dict

```
__________ TestWebEngineVersions.test_chromium_security_version_dict ___________

self = <test_version.TestWebEngineVersions object at 0x7ff5b9f9b8c0>
qapp = <PyQt6.QtWidgets.QApplication object at 0x7ff5c1759b30>

    def test_chromium_security_version_dict(self, qapp):
        """Check if we infer the QtWebEngine security version properly.
    
        Note this test mostly tests that our overview in version.py (also
        intended for human readers) is accurate. The code we call here is never
        going to be called in real-life situations, as the API is available.
        """
        try:
            from qutebrowser.qt.webenginecore import (
                qWebEngineVersion,
                qWebEngineChromiumSecurityPatchVersion,
            )
        except ImportError:
            pytest.skip("Requires QtWebEngine 6.3+")
    
        inferred = version.WebEngineVersions.from_webengine(
            qWebEngineVersion(), source="API")
>       assert inferred.chromium_security == qWebEngineChromiumSecurityPatchVersion()
Error: E       AssertionError: assert None == '122.0.6261.128'
E        +  where None = WebEngineVersions(webengine=VersionNumber(6, 8), chromium=None, source='API', chromium_security=None, chromium_major=None).chromium_security
E        +  and   '122.0.6261.128' = <built-in function qWebEngineChromiumSecurityPatchVersion>()
```

**Comment by The-Compiler** [comment]
# TestSavefileOpen.test_failing_flush failing

```
tests/unit/utils/test_qtutils.py:498: Failure: Qt messages with level WARNING or above emitted
----------------------------- Captured Qt messages -----------------------------
QtWarningMsg: QSaveFile::commit: File (/tmp/pytest-of-user/pytest-0/test_failing_flush0/foo) is not open
```

probably just a new warning from Qt we can suppress for that test?

**Comment by The-Compiler** [comment]
# Permission caching/saving breaks tests

```
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_async_question_interrupted_by_async_one - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_async_question_interrupted_by_blocking_one - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_geolocation_with_ask__false - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_geolocation_with_ask__false_and_save - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_geolocation_with_ask__abort - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_always_rejecting_notifications - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'js', 'message': '[*] notification permission denied'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_always_accepting_notifications - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'category': 'js', 'message': '[*] notification permission granted'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_notifications_with_ask__false - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_notifications_with_ask__false_and_save - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_notifications_with_ask__true - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_notifications_with_ask__true_and_save - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_answering_notification_after_closing_tab - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
Error: FAILED tests/end2end/features/test_prompts_bdd.py::test_interrupting_ssl_prompt_during_a_notification_prompt - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 15000ms waiting for {'message': 'Asking question *'}.
```

e.g.:

```
----> Waiting for 'Asking question *' in the log
[...]
  07:58:57.045 DEBUG    webelem    webelem:click:411 Clicking <qutebrowser.browser.webengine.webengineelem.WebEngineElement html='<input type="button" onclick="get_notification_permission()" value="Get notification permission" id="button">'> with click_target ClickTarget.normal, force_event False
[...]
  07:58:57.060 DEBUG    js         shared:javascript_log_message:190 [http://localhost:59211/data/prompt/notifications.html:28] [FAIL] unknown initial value for Notification.permission: granted
```

probably due to:

> Added [QWebEngineProfile::setPersistentPermissionsPolicy](https://doc-snapshots.qt.io/qt6-dev/qwebengineprofile.html#setPersistentPermissionsPolicy)() to control how website permissions persist across browsing sessions. By default, incoming permissions are saved to disk.

I suppose we'll need at least a `-D` flag (or maybe even a setting?) to set that to `NoPersistentPermissions` (?) for tests.

**Comment by The-Compiler** [comment]
Not much more we can do for Beta 1, I suppose:

- The Chromium base is still at 118, the 122-based upgrade should come with Beta 2. If we're unlucky, that'll come with some new regressions/issues too.
- We can only use [QWebEngineProfile::setPersistentPermissionsPolicy](https://doc-snapshots.qt.io/qt6-dev/qwebengineprofile.html#setPersistentPermissionsPolicy)() once we get an updated PyQt 6.8 with API for it

I guess we could still investigate the `QSaveFile` thing in the meantime. Beta 2 will be out on July 17th, though I'll be on holidays then and back on the 29th.

**Comment by toofar** [comment]
I've pushed some commits to silence the failing tests (not actual fixes) for now, and dug into the persistent permissions feature a tiny bit.

> Missing security patch version in our dict

I just set the version numbers to what Qt is reporting. Even though it is incorrect 1) the test is green now for people raising PRs 2) it'll go red again when things next change, hopefully to be more correct

> TestSavefileOpen.test_failing_flush failing

I just ignored the new warning message for that specific test. The test is deliberately trying to provoke this situations, so the warning is correct.

> Permission caching/saving breaks tests

This is the more complicated one. It's a new feature in webengine that can remember the feature permissions you grant or deny. By default it remembers the permissions you grant per origin both on disk and in memory. This differs from our current behaviour in that we don't remember selections in memory and we only save to disk when the user chooses to.

For now I've just skipped the failing tests, but I think we are going to have to do proper adaptations to this feature for the following reasons:

1. It's on by default, and as mentioned above will change our default behavior. Until PyQt updates to include the new 6.8 APIs (which can take anything from a week to a month or more) we cannot turn it off. Users are likely to have Qt6.8 with PyQt6.7 which means permissions grants will be remembered when they weren't before.
2. The persistence mechanism doesn't respect the APIs we use to separate basedirs, which means permission grants will leak between basedirs. Hopefully they fix this one, I've raised a bug ticket here: https://bugreports.qt.io/browse/QTBUG-126595

**Comment by toofar** [comment]
beta2 is out now and the only new failure I've seen was the "Third-party iframes in qutebrowser stylesheet script" test consistently failing. Seems it's actually working but the error message has change a bit. Hopefully that's all it is and we don't have other weird stuff with frames cropping up. https://github.com/qutebrowser/qutebrowser/commit/43fa657f55c08256c1f8fe39b9c1348f499f4831

There does seem to be some other tests which look a bit flaky, have to keep an eye on them.

**Comment by toofar** [comment]
Had my first Google Meet call on Qt 6.8 today. It crashes the browser again, yippee!

Stack trace:
```
#0  0x00007fffe5343320 in WebRtcLogUploader::ApplyForStartLogging() () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#1  0x00007fffe534f9b8 in WebRtcTextLogHandler::StartLogging(base::OnceCallback<void (bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#2  0x00007fffe5349f7d in WebRtcLoggingController::StartLogging(base::OnceCallback<void (bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#3  0x00007fffe5335ffe in extensions::WebrtcLoggingPrivateStartFunction::Run() () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#4  0x00007fffe8e753a4 in ExtensionFunction::RunWithValidation() () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#5  0x00007fffe8e785c5 in extensions::ExtensionFunctionDispatcher::DispatchWithCallbackInternal(extensions::mojom::RequestParams const&, content::RenderFrameHost*, content::RenderProcessHost&, base::OnceCallback<void (ExtensionFunction::ResponseType, base::Value::List, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mojo::StructPtr<extensions::mojom::ExtraResponseData>)>) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#6  0x00007fffe8e7933f in extensions::ExtensionFunctionDispatcher::Dispatch(mojo::StructPtr<extensions::mojom::RequestParams>, content::RenderFrameHost&, base::OnceCallback<void (bool, base::Value::List, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mojo::StructPtr<extensions::mojom::ExtraResponseData>)>) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#7  0x00007fffe8e686ed in extensions::ExtensionFrameHost::Request(mojo::StructPtr<extensions::mojom::RequestParams>, base::OnceCallback<void (bool, base::Value::List, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mojo::StructPtr<extensions::mojom::ExtraResponseData>)>) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#8  0x00007fffe8d589b3 in extensions::mojom::LocalFrameHostStubDispatch::AcceptWithResponder(extensions::mojom::LocalFrameHost*, mojo::Message*, std::unique_ptr<mojo::MessageReceiverWithStatus, std::default_delete<mojo::MessageReceiverWithStatus> >) ()
   from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#9  0x00007fffe8e6875d in extensions::mojom::LocalFrameHostStub<mojo::RawPtrImplRefTraits<extensions::mojom::LocalFrameHost> >::AcceptWithResponder(mojo::Message*, std::unique_ptr<mojo::MessageReceiverWithStatus, std::default_delete<mojo::MessageReceiverWithStatus> >) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#10 0x00007fffe9f1db95 in mojo::InterfaceEndpointClient::HandleValidatedMessage(mojo::Message*) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#11 0x00007fffe9f1ee37 in mojo::MessageDispatcher::Accept(mojo::Message*) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#12 0x00007fffe9f1d64e in mojo::InterfaceEndpointClient::HandleIncomingMessage(mojo::Message*) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#13 0x00007fffea17a18a in IPC::ChannelAssociatedGroupController::AcceptOnEndpointThread(mojo::Message, IPC::(anonymous namespace)::ScopedUrgentMessageNotification) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
#14 0x00007fffea176643 in base::internal::Invoker<base::internal::BindState<void (IPC::ChannelAssociatedGroupController::*)(mojo::Message, IPC::(anonymous namespace)::ScopedUrgentMessageNotification), scoped_refptr<IPC::ChannelAssociatedGroupController>, mojo::Message, IPC::(anonymous namespace)::ScopedUrgentMessageNotification>, void ()>::RunOnce(base::internal::BindStateBase*) () from /usr/local/Qt-6.8.0/lib/libQt6WebEngineCore.so.6
```

I'm on 3f4790463cc5 (beta2 plus a bit, not the latest commit).
Enabling the pakjoy workaround prevents the crash.
Can use the same reproducer as last time, load  https://meet.google.com/test and click the cog up the top right of the page.

**Comment by toofar** [comment]
### Drag and drop not working.

I'm still on an older commit (`3f4790463cc5` in the qt5 repo, I've tried building newer commits but I keep getting renerer process crashes (status 5), not sure if the 6.8 branch is broken or my build is) but I've found dragging elements around in a few sites freezes up the whole browser when you start drading and then releasing the element doesn't take effect. I've seen this in the wild on the outlook web app and can reproduce on https://codepen.io/Goldfsh/pen/zBbOqm

It seems to freeze the browser for 3s or so at a time while you are dragging and on stdout I see:

WARNING: WebContentsAdapter::updateDragAction was not called within 3000 ms.
WARNING: WebContentsAdapter::updateDragAction was not called within 3000 ms.
WARNING: WebContentsAdapter::updateDragAction was not called within 3000 ms.
WARNING: WebContentsAdapter::updateDragAction was not called within 3000 ms.

I'll try to repro in the kde-unstable docker container later to isolate the "is it my build" question.

**Comment by toofar** [comment]
A minor annoyance on the latest commit of the 6.8 branch (67dcf8ced943a). I've got the outlook web app in the first tab and use alt+1 to switch to it. It now picks up the "Alt" keypress and shows these mnemonic shortcut hints when it didn't used to (they aren't qutebrowser hints, I guess they are an accessibility feature of outlook). I guess I could bind Alt in normal mode to `nop`.

![image](https://github.com/user-attachments/assets/8a30e6f9-a82e-4077-96fc-a8f36ded65f4)

**Comment by toofar** [comment]
A test has started failing with beta4 because QWebEngineProfile.isSpellCheckEnabled() is reporting False where it used to be True https://github.com/qutebrowser/qutebrowser/actions/runs/10735316586/job/29772339738
I don't yet understand what the test is supposed to be doing or why it's changed. Might be something to do with the changes around dictionary search path error handling (https://bugreports.qt.io/browse/QTBUG-124110) but I don't understand whether something has actually broken or whether the test was never working. Where is the en-US dict installed? I straced that test on 6.7 and don't see any bdic files loaded. Has `setSpellCheckLanguages()` or `setSpellCheckEnabled()` working? Spell checking still seems to be working for me.

**Comment by toofar** [comment]
I think there's been a regression in the `content.javascript.clipboard` handling in that the `access` option no longer works. Not 100% sure, but I think what happened is that with the new feature permission it'll send `featurePermissionRequested` if you don't have full access configured (`access-paste`), and since we don't handle that it goes ahead and denies access. It's a bit hard to follow the permissions changes upstream through because it was spread over a bunch of changes and not linked to a ticket.

**Comment by toofar** [comment]
Ah, 6.8 is supposed to be releasing [on Tuesday](https://wiki.qt.io/Qt_6.8_Release).
I'll have a quick look at the tests and see if I can whip up something around the permissions based on the PyQt patch I've been using.

**Comment by The-Compiler** [comment]
After things were green with 7475d38 for a little, unstable CI now [fails again](https://github.com/qutebrowser/qutebrowser/actions/runs/11315533769/job/31466825083):

```pytb
ImportError while importing test module '/__w/qutebrowser/qutebrowser/tests/unit/misc/test_ipc.py'.

Traceback:
[...]
tests/unit/misc/test_ipc.py:20: in <module>
    from qutebrowser.qt.test import QSignalSpy
E   ImportError: cannot import name 'QSignalSpy' from 'qutebrowser.qt.test' (/__w/qutebrowser/qutebrowser/qutebrowser/qt/test.py)
```

This seems due to [Archlinux patching it out](https://gitlab.archlinux.org/archlinux/packaging/packages/pyqt6/-/commit/ca31ea9bbcd75d5fd3f79939f5cda14075acdc23#9bab99adb6bd9264c834cbf08c31fb3f3ec758c0_0_288) with their PyQt 6.8 build. According to the maintainer:

> yes, there are big changes in 6.8 that break the sip file shipped by pyqt 6.7. I'm not moving it out of testing until pyqt 6.8 is out or I figure out a fix
> help is welcome if course :)
> more specifically, qsignalspy is no longer derived from qobject in 6.8

Will push a fix that just downgrades to PyQt6 6.7.1-2 (from `[extra]`) for the moment, as we can't easily avoid using `QSignalSpy` (the `test_ipc.py` one is easy, but we also use it in `testprocess.py` to wait for lines).

**Comment by jpka-** [comment]
I confirm comment [above](https://github.com/qutebrowser/qutebrowser/issues/8242#issuecomment-2416423234). Qt 6.8 introduce regression (again) with bitmap fonts 1) are not displayed, and 2) can't be selected using `qt6ct` anymore (xos4 Terminus and VGA are vital). The only way to solve is downgrade qt6-* to 6.7.3, for Arch it's per [here](https://wiki.archlinux.org/title/Arch_Linux_Archive#How_to_restore_all_packages_to_a_specific_date) and last working archive date is 24/10/14.

**Comment by nwoeanhinnogaehr** [comment]
I have the same problem. Using "xos4 Terminus" as a workaround is extremely slow, with interface often freezing for a second at a time whenever text needs to be rendered. Profiling shows that the vast majority of cpu time is being spent in `libz.so` (particularly, the `inflate` method).

**Comment by bitraid** [comment]
OK, last post about this, to let everyone know that a fix has been merged today (though not a proper one):
https://codereview.qt-project.org/c/qt/qtbase/+/602168

**Comment by nathanielevan** [comment]
> Is this the place to complain about Qt 6.8 problems? Because I have a weird one:
> 
> It seems that there are (again) problems with bitmap fonts.

Hi, pitching in here to confirm that I have this same issue, same qutebrowser version and all. Tried with `--temp-basedir` flag and changed the font and the same thing happened -- the font seemed to be recognised but no text is showing on the UI at all. I suppose we'll wait and see

**Comment by The-Compiler** [comment]
Looking through the comment backlog here:

- Permission persistence: Fixed via #8348
- Clipboard `access` not working: Fixed in Qt 6.8.1: [[QTBUG-130599] [Regr: 6.7->6.8]JavascriptCanAccessClipboard doesn't allowing copying to clipboard on 6.8 - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-130599)
- `isSpellCheckEnabled()` issue: Tracked in #8330 
- Font rendering issue: Should be [fixed](https://codereview.qt-project.org/c/qt/qtbase/+/602257) with Qt 6.8.1
- Various small changed error messages and such: Fixed by @toofar in `main`
- Other issues during beta: Fixed upstream already

I think that only leaves us with the [Alt shortcut issue](https://github.com/qutebrowser/qutebrowser/issues/8242#issuecomment-2324101356) - @toofar is this still current?

**Comment by toofar** [comment]
> I think that only leaves us with the [Alt shortcut issue](https://github.com/qutebrowser/qutebrowser/issues/8242#issuecomment-2324101356) - @toofar is this still current?

Yeah, still happening on the webengine 6.8 branch from November 15 (`d661a4c7eddd7f`). I tried to look through the outlook code to see what kind of trigger it was using, but it's all minified framework code and I couldn't find anything. If anyone wants to figure out what changed, perhaps trying to make a test page to print the status of various javascript key event APIs would be the way to go :shrug:
